temporalio/activity.py

```diff
@@ -0,0 +1,287 @@
+    return not _current_context.get(None) is None
+
+
+def info() -> Info:
+    """Current activity's info.
+
+    Returns:
+        Info for the currently running activity.
+
+    Raises:
+        RuntimeError: When not in an activity.
+    """
+    return _Context.current().info()
+
+
+def heartbeat(*details: Any) -> None:
+    """Send a heartbeat for the current activity.
+
+    Raises:
+        RuntimeError: When not in an activity.
+    """
+    heartbeat_fn = _Context.current().heartbeat
+    if not heartbeat_fn:
+        raise RuntimeError("Can only execute heartbeat after interceptor init")
+    heartbeat_fn(*details)
+
+
+def cancelled() -> bool:
+    """Whether this activity is cancelled.
+
+    Returns:
+        True if the activity has been cancelled, False otherwise.
+
+    Raises:
+        RuntimeError: When not in an activity.
+    """
+    return _Context.current().cancelled_event.thread_event.is_set()
+
+
+async def wait_for_cancelled() -> None:
+    """Asynchronously wait for this activity to be cancelled.
+
+    Raises:
+        RuntimeError: When not in an async activity.
+    """
+    async_event = _Context.current().cancelled_event.async_event
+    if not async_event:
+        raise RuntimeError("not in async activity")
+    await async_event.wait()
+
+
+def wait_for_cancelled_sync(timeout: Optional[float] = None) -> None:
+    """Synchronously block while waiting for cancellation on this activity.
+
+    This is essentially a wrapper around :py:meth:`threading.Event.wait`.
+
+    Args:
+        timeout: Max amount of time to wait for cancellation.
+
+    Raises:
+        RuntimeError: When not in an activity.
+    """
+    _Context.current().cancelled_event.thread_event.wait(timeout)
+
+
+def worker_shutdown() -> bool:
+    """Whether shutdown has been invoked on the worker.
+
+    Returns:
+        True if shutdown has been called on the worker, False otherwise.
+
+    Raises:
+        RuntimeError: When not in an activity.
+    """
+    return _Context.current().worker_shutdown_event.thread_event.is_set()
+
+
+async def wait_for_worker_shutdown() -> None:
+    """Asynchronously wait for shutdown to be called on the worker.
+
+    Raises:
+        RuntimeError: When not in an async activity.
+    """
+    async_event = _Context.current().worker_shutdown_event.async_event
+    if not async_event:
+        raise RuntimeError("not in async activity")
+    await async_event.wait()
+
+
+def wait_for_worker_shutdown_sync(timeout: Optional[float] = None) -> None:
+    """Synchronously block while waiting for shutdown to be called on the
+    worker.
+
+    This is essentially a wrapper around :py:meth:`threading.Event.wait`.
+
+    Args:
+        timeout: Max amount of time to wait for shutdown to be called on the
+            worker.
+
+    Raises:
+        RuntimeError: When not in an activity.
+    """
+    _Context.current().worker_shutdown_event.thread_event.wait(timeout)
+
+
+def raise_complete_async() -> NoReturn:
```

Comment:
As a rule of thumb I'd rather reduce the API surface.

Reply:
Then I can hide the error class :-) That does make more sense actually at first glance.

Or do we want to even throw at all? Go wants you to return an error, but Java has `doNotCompleteOnReturn` that basically does complete-async whenever the method ends up returning.

Reply:
I'd vote for returning a special _non-error_ value from the function. Since it's really not an "error" to want to continue as new, and the Java approach feels too indirect.

That way a workflow author can just write
`return CompleteAsync()`

If it's felt that complicates the return types too much then I understand wanting to keep it an exception, but it doesn't seem like it would too much to me.

Reply:
Yeah, we definitely want people to be able to trust their types, so if we can't just set a "ignore response" flag like Java, I'm thinking throw is best.

Reply:
Removed the exception from public API